Mage.Server.Plugins/Mage.Player.Human/src/mage/player/human/HumanPlayer.java

```diff
@@ -828,35 +826,49 @@ public boolean chooseTargetAmount(Outcome outcome, TargetAmount target, Ability
 
             UUID responseId = getFixedResponseUUID(game);
             if (responseId != null) {
-                if (targets.containsKey(responseId)) {
-                    targets.remove(responseId);
-                    possibleTargets.add(responseId);
+                if (target.contains(responseId)) {
+                    target.remove(responseId);
                 } else if (possibleTargets.contains(responseId) && target.canTarget(abilityControllerId, responseId, source, game)) {
-                    MageObject targetObject = game.getObject(responseId);
-                    String targetString;
-                    if (targetObject != null) {
-                        targetString = targetObject.getIdName();
-                    } else {
-                        targetString = responseId.toString();
-                    }
-                    targets.put(responseId, targetString);
-                    possibleTargets.remove(responseId);
+                    target.addTarget(responseId, source, game);
                 }
             } else if (!required) {
                 break;
             }
         }
 
+        List<UUID> targets = target.getTargets();
         if (targets.isEmpty()) {
             return false;
         }
 
-        List<Integer> amountList = getMultiAmount(new ArrayList<>(targets.values()), 1, amount, MultiAmountType.DAMAGE, game);
+        List<String> targetStrings = new ArrayList<>(amount);
+        for (UUID targetId : targets) {
+            MageObject targetObject = game.getObject(targetId);
+            if (targetObject != null) {
+                StringBuilder sb = new StringBuilder();
+                sb.append(targetObject.getIdName());
+                sb.append(", ");
+                sb.append(targetObject.getPower().getValue());
+                sb.append('/');
+                sb.append(targetObject.getToughness().getValue());
+                targetStrings.add(sb.toString());
+            } else {
+                Player player = game.getPlayer(targetId);
+                if (player != null) {
+                    targetStrings.add(player.getName());
+                } else {
+                    targetStrings.add(targetId.toString());
+                }
+            }
+        }
+
+        List<Integer> amountList = getMultiAmount(targetStrings, 1, amount, MultiAmountType.DAMAGE, game);
         int i = 0;
-        for (UUID targetId : targets.keySet()) {
-            target.addTarget(targetId, amountList.get(i), source, game);
+        for (UUID targetId : targets) {
+            target.setTargetAmount(targetId, amountList.get(i), game);
             i++;
         }
+        target.clearRemainingAmount();
```

Comment:
It's a dirty hack -- `remainingAmount` and `chosen` status must be auto-calculated after each change (add or remove target). So if you catch wrong `remainingAmount` then there is must be wrong adding/remove code usage.

![shot_210213_091544](https://user-images.githubusercontent.com/8344157/107842435-85cd7780-6ddc-11eb-8bb1-f49a537f8e1b.png)
![shot_210213_091555](https://user-images.githubusercontent.com/8344157/107842437-882fd180-6ddc-11eb-88a2-a2bc2bd479cd.png)
![shot_210213_093329](https://user-images.githubusercontent.com/8344157/107842684-8cf58500-6dde-11eb-869d-7af03c62f491.png)

Reply:
I think `setTargetAmount` can be wrong. I'll look at Spellskite's code and try to fix/remove it.

![shot_210213_093759](https://user-images.githubusercontent.com/8344157/107842778-2f156d00-6ddf-11eb-9af3-0ddde846f6c0.png)
![shot_210213_093819](https://user-images.githubusercontent.com/8344157/107842784-389ed500-6ddf-11eb-9662-4be016ac455d.png)

Reply:
Well I'm adding the target without an amount and then setting the amount at the end.

https://github.com/weirddan455/mage/blob/890de9a659255c0325fbc75b63c92128c6ae0480/Mage.Server.Plugins/Mage.Player.Human/src/mage/player/human/HumanPlayer.java#L868

Above method doesn't update the remaining amount.  But yea, it is kind of a dirty hack.

Reply:
Any status report on this?

Reply:
Yes, in the work